pyro/infer/rws.py

```diff
@@ -0,0 +1,242 @@
+import math
+
+import torch
+
+import pyro
+import pyro.poutine as poutine
+from pyro.infer.elbo import ELBO
+from pyro.infer.enum import get_importance_trace
+from pyro.infer.util import is_validation_enabled
+from pyro.poutine.util import prune_subsample_sites
+from pyro.util import check_if_enumerated, check_model_guide_match, warn_if_nan
+
+
+class ReweightedWakeSleep(ELBO):
+    r"""
+    An implementation of Reweighted Wake Sleep following reference [1].
+
+    .. note:: Sampling and log_prob evaluation asymptotic complexity:
+
+    1) Using wake-theta and/or wake-phi
+        O(`num_particles`) samples from guide,
+        O(`num_particles`) `log_prob` evaluations of model and guide
+
+    2) Using sleep-phi
+        O(`num_sleep_particles`) samples from model,
+        O(`num_sleep_particles`) `log_prob` evaluations of guide
+
+    if 1) and 2) are combined,
+        O(`num_particles`) samples from the guide,
+        O(`num_sleep_particles`) from the model,
+        O(`num_particles` + `num_sleep_particles`) `log_prob` evaluations of the guide, and
+        O(`num_particles`) evaluations of the model
+
+
+    .. note:: This is particularly useful for models with stochastic branching,
+        as described in [2].
+
+    .. note:: This returns _two_ losses, one each for the model and the guide.
```

Comment:
Could you mention the correspondence theta=model, phi=guide here? Or even go farther and use variable names like `wake_model_loss` and `guide_loss` in the code (only if you feel that does not stray too far from your paper).

Reply:
Done. Added correspondence (60486ebd)